Application Insights: Add Favorites and Web Test Locations APIs

d3r3kk opened the pull request:
__Web Test Locations API:__
Application Insights has web tests already defined in their REST API, but the locations for web tests were not discoverable for users at the time we wrote that spec. This PR brings the web test location query to our customers to make this information available.

__Favorites API:__
Application Insights components contain many reportable items and each of those items are highly configurable. We've given our users the ability to store their commonly used queries and reports as favorites within the Application Insights component. This PR exposes the management of these favorites to the REST API.

__Misc:__
Correct some minor testing failures with webTest and components APIs.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits...
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

d3r3kk committed:
Initial addition of favorites and locations API

d3r3kk committed:
Complete web test locations API

- favor 'web test' names/descriptions over 'synthetic monitor' as web test is more customer-centric naming
- create example data for the web test api

d3r3kk committed:
Spelling update

d3r3kk committed:
Updated parameters for favorite API

d3r3kk committed:
Add favorites API and examples

- ensure tests run against the applicationinsights folder successfully

d3r3kk committed:
Add collectionFormat to comma-separated tags list parameter for favorites
- generating API in CSharp generated errors

AutorestCI commented:
Did a commit to Azure/azure-sdk-for-python:
https://github.com/Azure/azure-sdk-for-python/commit/c184e9cc61828b665753770cdc9190346f5490fd

AutorestCI commented:
Did a commit to Azure/azure-sdk-for-go:
https://github.com/Azure/azure-sdk-for-go/commit/f05b00fd905c2be3ce125760c2eba627684fa9b5

olydis commented:
@ravbhatnagar new APIs

ravbhatnagar commented on specification/applicationinsights/resource-manager/microsoft.insights/stable/2015-05-01/favorites_API.json:
```diff
@@ -0,0 +1,403 @@
+    "ApplicationInsightsComponentFavorite": {
+      "description": "Properties that define a favorite that is associated to an Application Insights component.",
+      "properties": {
+        "Name": {
```
camel case all property names

d3r3kk replied:
The service I am defining via Swagger returns “PascalCase” property names and not “camelCase” as defined in our Swagger guidelines.

I can convert the swagger document property names to camelCase but I believe that breaks the auto-generated C# code?

johanste replied:
Whatever the outcome of the API design discussion, the swagger *must* be true to what the service actually returns. If the service is using PascalCase, then you should add suppressions to tell the linter that you are indeed doing this on purpose. This particular guidelines is about the API design, not about the swagger describing the service :).

d3r3kk replied:
Thanks for the feedback @johanste, I will add the necessary linter suppressions.

d3r3kk replied:
Done. I updated suppressions that were already there.

ravbhatnagar commented on specification/applicationinsights/resource-manager/microsoft.insights/stable/2015-05-01/favorites_API.json:
```diff
@@ -0,0 +1,403 @@
+          "type": "string",
+          "description": "The user-defined name of the favorite."
+        },
+        "Config": {
```
only certain top level properties are allowed by ARM. All RP specific resource properties should go inside the properties envelope. Is this new API or something which already exists on the service side?

zakimaksyutov replied:
This is not a new API, it already exists.

d3r3kk replied:
I believe I understand this. The ARM API requires things to be structured this way (CDS service returns properties that must be encased within the 'ARM' top-level properties).

However, our API exists outside of this structure, and returns these top level structures today and does not return the top-level envelope that ARM provides. Does this mean we do not require the top-level ARM envelope?

For instance, I see that our 'Components' API includes in its "definitions" the top-level 'Resource' container, but I don't see it being used anywhere there. Is this the requirement then, to define it but not use it in this case?

ravbhatnagar commented on specification/applicationinsights/resource-manager/microsoft.insights/stable/2015-05-01/favorites_API.json:
```diff
@@ -0,0 +1,403 @@
+  "definitions": {
+    "ApplicationInsightsComponentFavorite": {
+      "description": "Properties that define a favorite that is associated to an Application Insights component.",
+      "properties": {
```
id and type top level properties missing

ravbhatnagar commented:
Lets set up a call to understand this and see if would be correct to fix them considering these are existing. 
Please set up a call and loop in folks from Azure API review board as well - @johanste  @devigned  @NiklasGustafsson. Also please bucket all the AI APIs that have similar issues and we can discuss them together. I think there is another PR for AI which have these issues.

d3r3kk commented:
@ravbhatnagar agreed, I would like to ensure that we get this resolved very soon too! I'll reach out via Outlook and try and get a meeting in place shortly.

devigned commented on specification/applicationinsights/resource-manager/microsoft.insights/stable/2015-05-01/favorites_API.json:
```diff
@@ -0,0 +1,403 @@
+          ],
+          "x-ms-enum": {
+            "name": "FavoriteType",
+            "modelAsString": false
```
If an enum will ever have additional values, it should be modeled as string. If it will only have these values _forever_, then modelAsString false is appropriate. Please use this guidance for each of the enums.

d3r3kk replied:
I will investigate the field and determine if it will ever be extended and alter it to `"modelAsString": true` if this is the case.

d3r3kk replied:
Changed to modelAsString as we may need to add other favorite types in the future.

olydis replied:
@d3r3kk it sounds like you meant to set this to `true`? 🙂

ravbhatnagar commented:
Signing off from ARM side. Although there are compliance issues here, these APIs have been in prod for several years. Ongoing conversation with AI team on whether its possible to fix these in a new api-version.

lmazuel commented:
@AutorestCI rebuild azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-python

d3r3kk committed:
Enum for favorite type should be set to modelAsString as further favorite types are possible in future.

d3r3kk committed:
Update suppressions to include specific instances I need suppressed (no global suppressions where possible).

d3r3kk committed:
Suppress some noisey test warnings from API specifications from my peers.

AutorestCI commented:
# Automation for azure-sdk-for-go

d3r3kk commented:
Completed my updates. Could I get a review sign off from @johanste , @devigned , or @NiklasGustafsson ?

Note that I added 3 suppressions for noisy warnings from files not from my PR. (description/title missing and response schema - all warnings for redundant or false negative failures).

Thanks!

devigned commented:
@olydis any other feedback you'd like to impart prior to merging?

olydis commented on specification/applicationinsights/resource-manager/microsoft.insights/stable/2015-05-01/favorites_API.json:
```diff
@@ -0,0 +1,403 @@
+            "$ref": "./examples/FavoritesList.json"
+          }
+        },
+        "x-ms-pageable": {
```
the declared response type is not of pageable shape (object with `value` property of type array like in [this random example](https://github.com/Azure/azure-rest-api-specs-pr/blob/0bef2b99ad6d52781e3c308bc6b420f6eef59ef7/specification/resources/resource-manager/Microsoft.Resources/stable/2015-11-01/subscriptions.json#L181-L189) - what you have is a bare array), i.e. either the `schema` is wrong or this annotation 🙂

d3r3kk replied:
I checked the code in our service, but it never returns a paged item - only returns the entire list of favorites without chunking them into pages.
Removed the x-ms-pageable annotation.

I imagine this is also part of how we may be out of compliance with the ARM schemas at the moment (we are aware of this and future updates to the service are slated to bring us back into compliance, hence the back and forth with the ARM team earlier).

olydis commented on specification/applicationinsights/resource-manager/microsoft.insights/stable/2015-05-01/favorites_API.json:
```diff
@@ -0,0 +1,403 @@
+        "shared",
+        "user"
+      ],
+      "description": "The type of favorite. Value can be either shared or user.",
```
`x-ms-enum` missing, also please don't duplicate the `enum` information in the description 🙂

d3r3kk replied:
Only SourceTypeParameter needs to be modelled as a string, not FavoriteTypeParameter (sorry for any confusion on past comments).

Also, I added the `x-ms-enum` field in the missing instance.

olydis replied:
no worries, thanks for clarifying 🙂

olydis reviewed (changes requested):
just some minor comments, looking good overall!

d3r3kk committed:
PR suggestions

- Remove x-ms-pageable annotation from our list operation, in the service code we never paganate our data.
- Add x-ms-enum annotation to one enum data type where it was missing.
